ssl/quic/quic_impl.c

```diff
@@ -9,271 +9,1241 @@
 
 #include <openssl/macros.h>
 #include <openssl/objects.h>
+#include <openssl/sslerr.h>
+#include <crypto/rand.h>
 #include "quic_local.h"
+#include "internal/quic_dummy_handshake.h"
+#include "internal/quic_rx_depack.h"
+#include "internal/quic_error.h"
+#include "internal/time.h"
 
+static void aon_write_finish(QUIC_CONNECTION *qc);
+
+/*
+ * QUIC Front-End I/O API: Common Utilities
+ * ========================================
+ */
+
+/*
+ * Block until a predicate is met.
+ *
+ * Precondition: Must have a channel.
+ */
+static int block_until_pred(QUIC_CONNECTION *qc,
+                            int (*pred)(void *arg), void *pred_arg,
+                            uint32_t flags)
+{
+    QUIC_REACTOR *rtor;
+
+    assert(qc->ch != NULL);
+
+    rtor = ossl_quic_channel_get_reactor(qc->ch);
+    return ossl_quic_reactor_block_until_pred(rtor, pred, pred_arg, flags);
+}
+
+/*
+ * Raise a 'normal' error, meaning one that can be reported via SSL_get_error()
+ * rather than via ERR.
+ */
+static int quic_raise_normal_error(QUIC_CONNECTION *qc,
+                                   int err)
+{
+    qc->last_error = err;
+    return 0;
+}
+
+/*
+ * Raise a 'non-normal' error, meaning any error that is not reported via
+ * SSL_get_error() and must be reported via ERR.
+ */
+static int quic_raise_non_normal_error(QUIC_CONNECTION *qc,
+                                       const char *file,
+                                       int line,
+                                       const char *func,
+                                       int reason,
+                                       const char *fmt,
+                                       ...)
+{
+    va_list args;
+
+    ERR_new();
+    ERR_set_debug(file, line, func);
+
+    va_start(args, fmt);
+    ERR_vset_error(ERR_LIB_SSL, reason, fmt, args);
+    va_end(args);
+
+    qc->last_error = SSL_ERROR_SSL;
+    return 0;
+}
+
+#define QUIC_RAISE_NORMAL_ERROR(qc, err)                        \
+    quic_raise_normal_error((qc), (err))
+
+#define QUIC_RAISE_NON_NORMAL_ERROR(qc, reason, msg)            \
+    quic_raise_non_normal_error((qc),                           \
+                                OPENSSL_FILE, OPENSSL_LINE,     \
+                                OPENSSL_FUNC,                   \
+                                (reason),                       \
+                                (msg))
+
+/*
+ * Should be called at entry of every public function to confirm we have a valid
+ * QUIC_CONNECTION.
+ */
+static ossl_inline int expect_quic_conn(const QUIC_CONNECTION *qc)
+{
+    if (!ossl_assert(qc != NULL))
+        return QUIC_RAISE_NON_NORMAL_ERROR(NULL, ERR_R_INTERNAL_ERROR, NULL);
```

Comment:
IMO this is wrong ERR reason. I'd add something like SSL_R_NOT_A_QUIC_CONNECTION. Also raising it here instead of directly at the place where qc is obtained from SSL object means the FILE and LINE information will be lost.

Reply:
A call to an SSL function should never reach these functions in the first place if it's not a QUIC connection, though?

Reply:
OK then.